mindmeld/components/dialogue.py

```diff
@@ -689,6 +692,189 @@ def _get_dialogue_handler(self, dialogue_state):
+
+        """ Extracting the Query object associated with the user input here.
+
+        process_markup args:
+            text, query factory, query options (optional dict) {language, time_zone and time_stamp}
+
+        returns:
+            tuple(raw_text (str), Query object, list of entities)
+
+        The following returns the extracted Query object
+        """
+        query = process_markup(text, resource_loader.query_factory, {})[1]
+
+        return query
+
+    def _validate(self, text, entity_type, slot):
+        """Validates the user input based on the entity type and validation type.
+
+        Args:
+            text (Request.text): Text in the query
+            entity_type: Type of entity
+            slot: FormEntity object
+
+        Returns:
+            (bool): Boolean whether the user input fulfills the slot requirements
+        """
+
+        query = self._extract_query_features(text)
+
+        """
+        payload format for entity feature extractors:
+            tuple(query (Query Object), list of entities, entity index)
+
+        For slot filling, the user query will consist of the prompted entity.
+        Hence `entity = [query]` and `entity_index = 0`.
+
+        """
+        formatted_payload = (query, [query], 0)
+
+        if slot.default_eval:
+            if entity_type in DEFAULT_SYS_ENTITIES:
+                # system entity validation - checks for presence of required system entity
+                resources = {}
+                extracted_feature = dict(
+                    query_features.extract_sys_candidates([entity_type])(query, resources)
+                )
+            else:
+                # gazetteer validation
+                gazetteer = (
+                    {'gazetteers':
+                        {entity_type:
+                            self._app.app_manager.nlp.resource_loader.get_gazetteer(entity_type)
+                        }
+                    }
+                )
+                extracted_feature = (
+                entity_features.extract_in_gaz_features()(formatted_payload, gazetteer))
+
+        if slot.hints:
+            # hints / user-list validation
+            extracted_feature = {'hint_validated_entity': text} if text in slot.hints else {}
+
+        # return True iff user input results in extracted features (i.e. validated)
+        return extracted_feature != {}
+
+    def _initial_fill(self, request):
+        """
+        Performs the first pass and fills the entity form with entity values available
+        in the initial query.
+
+        Args:
+            request (Request): The request object.
+        """
+        for entity in request.entities:
+            entity_type = entity['type']
+            role = entity['role']
+            value = entity['text']
+
+            for slot in self._local_form:
+                if entity_type == slot.entity:
+                    if slot.role == None or (role == slot.role):
+                        slot.value = value
```

Comment:
we should break the loop here since we only expect one slot in the local_form for each entity

Reply:
I can think of usecases where the same entity might fill two slots?

Reply:
did we decide to break here?

Reply:
I agree with ritvik, we should not break here since we should fill as many slots as possible